resholve: 0.6.0 -> 0.6.6, add resholveScript* fns

abathur opened the pull request:
###### Motivation for this change

This tracks a few upstream changes in resholve and dependencies:
1. extract argument-handling functions from `resholvePackage` to enable easier re-use in additional APIs...
2. address minor review feedback in the course of #138080 
3. add `resholveScript` and `resholveScriptBin` functions to close #138442

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

abathur committed:
resholvePackage: extract util functions

Extract argument-handling utility functions to prepare for adding
resholveScript* functions.

This tracks upstream work, but I broke it up a little more semantically here
in case it aids review. See:
https://github.com/abathur/resholve/commit/6aab74820531673c760d95c24b28f0c29f43fe01

abathur committed:
resholve: fix review nits from #138080

abathur committed:
resholve: 0.6.0 -> 0.6.1, add resholveScript* fns

r-rmcgibbo commented:
Result of `nixpkgs-review pr 139253` at d88181ba run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built successfully:</summary>
  <ul>
    <li>arch-install-scripts</li>
    <li>bashup-events32</li>
    <li>bashup-events44</li>
    <li>dgoss</li>
    <li>resholve</li>
  </ul>
</details>

happysalada commented:
Hey, I was just reading the whole doc again to make sure I understood and found a couple of typos.
This doesn't need to be handled in that PR, but I thought you might want to know
`convienience.` -> `convenience`
`isnert` -> `insert`

you have the following paragraph
```
# --fake 'f:setUp;tearDown builtin:setopt source:/etc/bashrc'
fake = {
  # fake accepts the initial of valid identifier types as a CLI convienience.
  # Use full names in the Nix API.
  function = [ "setUp" "tearDown" ];
  builtin = [ "setopt" ];
  source = [ "/etc/bashrc" ];
};
```
shouldn't the flag to `--fake` be `function` instead of just `f:` ?

in the following paragraph
```
# --fix 'aliases xargs:ls $GIT:gix'
fix = {
  # all single-word directives use `true` as value
  aliases = true;
  xargs = [ "ls" ];
  "$GIT" = [ "gix" ];
};
```
I still can't figure out what the `xargs = [ "ls" ];` does.

in the following paragraph
```
# --keep 'which:git;ls .:$HOME $LS:exa /etc/bashrc ~/.bashrc'
keep = {
  which = [ "git" "ls" ];
  "." = [ "$HOME" ];
  "$LS" = [ "exa" ];
  "/etc/bashrc" = true;
  "~/.bashrc" = true;
};
```
I still can't understand the following lines
`"." = [ "$HOME" ];`
` "$LS" = [ "exa" ];` shouldn't that line be in the fix section ?

The doc is otherwise amazing!

I understand these comments don't pertain to the current PR, so feel free to ignore them if you don't have time.

abathur commented:
> I understand these comments don't pertain to the current PR, so feel free to ignore them if you don't have time.

No worries. I'm leery of picking at the trivial stuff atm not because I don't think it's worth fixing--I prefer fixing them before I have a chance to forget--but mostly because the ofborg darwin checks are slow atm.

> This doesn't need to be handled in that PR, but I thought you might want to know
> `convienience.` -> `convenience`
> `isnert` -> `insert`

Fixed locally. 

A lot of the remaining parts hinge on information in the manpage. resholve is still lacking some important automation (particularly wrt single-sourcing the docs and maintaining the Nix expressions on both ends). I'd like to have ~full docs for the Nix and CLI APIs, but for now I'm offloading as much as I feel like I can get by with on the manpage.

I've added a local note to make the necessity of the manpage more evident.

> ...
> shouldn't the flag to `--fake` be `function` instead of just `f:` ?

No, but it's a fair question! In the Nix API I generally model verbose-but-explicit forms, but there's also an abbreviated version of the fake forms supported at the CLI. The doc here is trying to bridge between the two APIs by demonstrating how the terse manpage form translates to the verbose Nix form. But yeah, you need both to decode it. 

(If anyone reading this has a handle on patterns for granular single-sourcing like this, I have an issue open at abathur/resholve#19).

> ...
> I still can't figure out what the `xargs = [ "ls" ];` does.
> ...
> I still can't understand the following lines
> `"." = [ "$HOME" ];`
> ` "$LS" = [ "exa" ];` shouldn't that line be in the fix section ?

I can explain these (and don't mind doing so!) but I am curious if the manpage clarifies them for you (or if you've already read it and they were still unclear.) Don't feel obliged, though.

abathur committed:
resholve: 0.6.1 -> 0.6.2

happysalada commented:
Thank you for your answers!

For other people who are following, here is a link to the man page https://github.com/abathur/resholve/blob/master/resholve.1.txt

The relevant section for `--keep` says
```
    −−keep ’directives’ | RESHOLVE_KEEP=’directives’
         Ignore (omit errors for) some things it isn’t safe to fix/ignore by
         default. Directive forms:

         outer_command:inner_command[;inner_command]...
            ignore inner_command(s) in the arguments to outer_command

         $variable (Note: $ may need escaping)
            ignore $variable or ${variable} used as all or part of a first
            word (command)

            Such as: ‘$GIT status‘ or ‘$BUSYBOX/ls‘

         absolute_path
            ignore a specific absolute_path

         ~/path
            ignore a specific home‐relative_path

         Ex: −−keep ’which:git;ls source:$HOME $LS:exa /etc/bashrc ~/.bashrc’
```

reading that I'm guessing that the line
`"$LS" = [ "exa" ];` means ignore the error for the variable `$LS` that part is clear. reading from the doc, perhaps the second part means that you ignore it when it's used like `$LS exa` ? This is the part that doesn't make sense, I couldn't figure out why you would write such a command.

then the
`"." = [ "$HOME" ];` would mean ignore the `.` path relative to `$HOME` ? (not sure about this one). It is one of the two paths possibilities, however not sure which one.

abathur committed:
resholve: 0.6.2 -> 0.6.3, fix readme

abathur commented:
Oof. Thanks for pulling on this thread--it flushed out a few inconsistencies here (which is frustrating, since this exactly what I'm hoping to avoid by not fully-duplicating documentation until I can get it single-sourced...)

From your previous post:
> I still can't figure out what the `xargs = [ "ls" ];` does.

And from this one:
> reading that I'm guessing that the line `"$LS" = [ "exa" ];` means ignore the error for the variable `$LS` that part is clear. reading from the doc, perhaps the second part means that you ignore it when it's used like `$LS exa` ? This is the part that doesn't make sense, I couldn't figure out why you would write such a command.

Both of these forms are syntactically valid but unimplemented. They leaked into the examples from tentative syntax forms that are commented out in the manpage source. 
<details><summary>(long story:</summary> I was having trouble settling on a CLI API. One way I sorted out my thoughts was by writing a manpage that included all of the options I imagined it gaining over the next few years to make sure there was conceptual/grammatical room. Then I came back and commented out what wasn't implemented, but obviously missed some forms in the examples :/)</details>

> then the `"." = [ "$HOME" ];` would mean ignore the `.` path relative to `$HOME` ?

If I'm reading you right, this is roughly correct. Technically, it's just telling resholve not to object to `$HOME` in this argument, without regard for whether it is the whole path, a prefix, in the middle, or a suffix.

Also, there's another good catch here. I missed updating the `"."` form here to track API changes that did get updated in the manpage. resholve now internally normalizes instances of `.` to `source`, so `"source" = [ "$HOME" ];` or `source = [ "$HOME" ];` are now the only sensical forms here.

I was wrapping up a bug-fix when you posted this, so I went ahead and made these doc fixes and and included them in the push.

abathur committed:
resholve: 0.6.3 -> 0.6.4

happysalada commented on pkgs/development/misc/resholve/README.md:
```diff
@@ -135,31 +161,32 @@ from the manpage, and the Nix equivalents:
-  which = [ "git" "ls" ];
-  "." = [ "$HOME" ];
+  source = [ "$HOME" ];
   "$LS" = [ "exa" ];
```
did you mean to remove that line as well ?

abathur replied:
Grumble. :)

abathur committed:
resholve: 0.6.4 -> 0.6.5

abathur committed:
resholve: 0.6.5 -> 0.6.6, respect buildInputs

abathur commented:
For clarity, since I keep adding commits to this PR: this PR is ready to be merged at any time.

I'm just tacking on new resholve releases as long as it stays open.

happysalada commented:
Then let's merge, ti think this adds hugely valuable features!
Thank you for the hard work!

happysalada commented:
I forgot to say, but feel free to tag me on your next PR, I really like this project, happy to review.
